### ISSUE TITLE
No state parameter specified in Linkedin Oauth

### ISSUE THREAD

**Issue by jgentes** [sahat/satellizer#267]
Hi, I'm struggling with an IE issue (oauth doesn't pass code from popup back to parent window) and trying to go back to basics for testing purposes.

I've setup your example app using Node on the backend. Google oauth works fine, but Linkedin fails with this error:

?error=invalid_request&error_description=You+need+to+pass+the+%22state%22+parameter

I believe everything is setup properly (not much to configure, really) - you can see it here: http://jg-sat-test.herokuapp.com/#/login

And the code is here:

https://ide.c9.io/jgentes/sat

**Comment by jgentes** [comment]
I just confirmed that now happens with my production install after upgrading the *bower* install of satellizer to .9.0. It was working fine before the upgrade.

The good news is that the IE window now closes, and both Chrome/IE have the same issue of not passing the STATE. I'll keep looking into it.

**Comment by jgentes** [comment]
Ok, I know this is a complete hack, but I needed a fix today and I am very happy to say that it has fixed the IE problem I was having, which was primarily due to having an old version of satellite in my bower.json file (vs. the new version in my package.json file for node)

Here's what I hacked to fix it, I hope it helps you figure out the problem:

I forced `+ '&state=STATE'` into the querystring:

`var url = defaults.authorizationEndpoint + '?' + oauth2.buildQueryString() + '&state=STATE';`

Then I commented out the check for matching state in localstorage:
`
            return popup.open(url, defaults.popupOptions)
              .then(function(oauthData) {
                if (defaults.responseType === 'token') {
                  return oauthData;
                }
                /*
                if (oauthData.state && oauthData.state !== $window.localStorage[stateName]) {
                  return $q.reject('OAuth 2.0 state parameter mismatch.');
                }
                */
                return oauth2.exchangeForToken(oauthData, userData);
              });
`

**Comment by sahat** [comment]
I will take a look at it today. Although, I already know what is the issue. In v0.9 I have added support for OAuth 2.0 state. Previously, LinkedIn had a hardcoded `state` string by default which is probably no longer being sent because it is overriden by the generic state function for OAuth 2.0 providers.

**Comment by sahat** [comment]
@jgentes I just double checked the code and it looks good to me.

It turns out that for LinkedIn as you as you specified `requiredUrlParams: ['state']`, which it is by default, it does not matter whether you set `state` or not, it defaults to `STATE` (even if you remove `state: 'STATE'` from *satellizer.js*.

![screenshot 2015-01-31 17 14 36](https://cloud.githubusercontent.com/assets/544954/5990606/f44c91cc-a96c-11e4-9afd-56bf5ae3e263.png)

I have only tested this with Firefox and Chrome, but I would imagine IE would be no different.

**Comment by jgentes** [comment]
Thank you for testing that.

The Heroku install I setup was using the example code you provided with no customization. Any idea why that's not working? Does it work for you when you test 
http://jg-sat-test.herokuapp.com/#/

-James

**Comment by sahat** [comment]
What would I do without Google Chrome debugger. I was able to debug the problem on the spot:

![screenshot 2015-01-31 21 16 07](https://cloud.githubusercontent.com/assets/544954/5991121/7f1688c8-a98e-11e4-8210-9f623fa49867.png)

Find this line in this OAuth 2.0 service:

```js
if (angular.isFunction(defaults.state)) {
  $window.localStorage[stateName] = defaults.state();
}
```

and change it to:

```js
if (angular.isFunction(defaults.state)) {
  $window.localStorage[stateName] = defaults.state();
} else if (angular.isString(defaults.state)) {
  $window.localStorage[stateName] = defaults.state;
}
```

I have already made that change, but will push it out and bump the version tomorrow.

**Comment by jgentes** [comment]
That's great, thank you - I appreciate all the time you put into this. It
has saved me a ton!

**Comment by sahat** [comment]
Sure. Fixed in 0.9.2.
